src/backend/distributed/connection/connection_management.c

```diff
@@ -492,6 +492,12 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 			continue;
 		}
 
+		if (RemoteConnectionClosed(connection))
+		{
+			elog(DEBUG1, "Skipping connection because it is closed");
```

Comment:
a lot of things could have gone wrong here. Should we be doing something along the lines of CheckRemoteTransactionsHealth?

Reply:
The problem is that other than `WL_SOCKET_CLOSED` we cannot deduce whether the remote socket is closed or not -- unless we try to read/write to the socket. I think we can consider doing something like CheckRemoteTransactionsHealth in addition to `WL_SOCKET_CLOSED` check.  Are you suggesting that or completely getting rid of `WL_SOCKET_CLOSED`?


You can also see some discussions here: https://github.com/postgres/postgres/commit/50e570a59e7f86bb41f029a66b781fc79b8d50f1

According to the discussions pgpool seems to send a remote command to check this.

Reply:
I guess what I'm trying to say is: it makes no sense to me to simply skip over the connection just because it is closed. What if we did something essential on it? We should be erroring out in that case, but then this does not seem like the right function to error out in. So perhaps we should check the health of our connections before FindAvailableConnection, remove/disable ones that are unessential, and error out if there are essential ones.

Reply:
Although this suggestion makes a lot of sense, I think it cases some error messages to become less informative:

> SELECT isolate_tenant_to_new_shard('table_1', 5, 'CASCADE', shard_transfer_mode := 'force_logical');
-ERROR:  connection not open
-CONTEXT:  while executing command on localhost:xxxxx
+ERROR:  failure on connection marked as essential: localhost:9060


The problem is that before this change, we'd be letting an unhealthy connection to execute the command. Now, we throw error earlier.

And, all these error message changes are exclusively happening for tenant isolation. Am I missing any details regarding tenant isolation on `EnsureAndGetHealthyConnections` function? Do we have any specific logic that I should be careful here? See all the CI output diffs: https://github.com/citusdata/citus/pull/6259/commits/c281a819e56feadbd94132cc7d71ffe52e2bb3db


Especially the cancellation failure error messages are interesting:
```
SELECT isolate_tenant_to_new_shard('table_1', 5, 'CASCADE', shard_transfer_mode := 'force_logical');
-ERROR:  canceling statement due to user request
+ERROR:  connection to the remote node localhost:9060 failed
```

Reply:
Ok, the specific logic for tenant isolation is because of PG_TRY/PG_CATCH. The function I wrote seems should be a little careful for future `OptionalCommandExecution` calls. Basically, allow OUTSIDE-TRANSCTION flag to skip error checks, because just in case, it might finish -- as the tests show.